[CL-1009-B] Tests for useHomepageSettingsFeatureFlag

lilkraftwerk opened the pull request:
Double check the actual response from the API but in my testing in the other branch I believe the `data` and `attributes` keys are necessary for the `useHomepageSettings` hook

lilkraftwerk committed:
Add test for UseHomepageSettingsFeatureFlag

lilkraftwerk committed:
Additional useHomepageSettingsFeatureFlag tests

brentguf commented on front/app/hooks/useHomepageSettingsFeatureFlag.ts:
```diff
@@ -40,7 +40,7 @@ export default function useHomepageSettingsFeatureFlag({
       : null;
   const isEnabled = !isNilOrError(homepageSettings)
-    ? homepageSettings.attributes[homepageEnabledSetting]
+    ? homepageSettings.data.attributes[homepageEnabledSetting]
```
We can also return the content of data, so we don’t need to do the .data each time (eg useIdea.ts).

lilkraftwerk replied:
+1, though this change will break your branch (I think) and my work on 1010 + 1011 so it might be worth doing later as a refactor

brentguf commented on front/app/hooks/useHomepageSettingsFeatureFlag.test.ts:
```diff
@@ -0,0 +1,62 @@
+import { renderHook } from '@testing-library/react-hooks';
```
Great! Can you also add a test for the case we have a section that doesn't need checking in appConfig?

lilkraftwerk replied:
edit: I understand, on it

brentguf replied:
Just clarifying so we're 💯  on the same page: a test for e.g. the top_info section (that just needs the enabled check in homepageSettings, not the allowed check in appConfiguration).

lilkraftwerk replied:
Just pushed that

brentguf commented on front/app/services/homepageSettings.ts:
```diff
@@ -26,7 +26,7 @@ export interface IHomepageSettings {
   // To check
   type: string;
-  attributes: IHomepageSettingsAttributes;
+  data: { attributes: IHomepageSettingsAttributes };
```
We usually have a Data type then, as you can see in most services files (eg IIdeaData in ideas.ts).

lilkraftwerk replied:
Something like this?

```
export interface IHomepageSettings {
  // To check
  id: string;
  // To check
  type: string;
  data: IHomepageSettingsData;
}

export interface IHomepageSettingsData {
  attributes: IHomepageSettingsAttributes;
}
```

brentguf replied:
Yes!

lilkraftwerk committed:
Add additional test and homepageSettings data type

lilkraftwerk committed:
Merge branch 'CL-1009/homepages-front-end-data' of github.com:CitizenLabDotCo/citizenlab into CL-1009/homepages-front-end-data-B

lilkraftwerk committed:
Update test for homepageSetting without appConfig

brentguf commented:
Could you solve merge conflicts and reassign me?

lilkraftwerk committed:
Merge branch 'CL-1009/homepages-front-end-data' of github.com:CitizenLabDotCo/citizenlab into CL-1009/homepages-front-end-data-B

lilkraftwerk committed:
Add more tests and clean up ternary operators

lilkraftwerk commented:
Added a couple more tests for just homepageSettings logic, and I found the boolean logic in the original hook a little dense so so I renamed some of the variables and expanded some of the ternary operators into additional lines

lilkraftwerk committed:
Update comment

brentguf commented on front/app/hooks/useHomepageSettingsFeatureFlag.ts:
```diff
@@ -31,19 +31,34 @@ export default function useHomepageSettingsFeatureFlag({
+    ? appConfigSetting.allowed
+    : false;
+
+  return isAllowedInAppConfig && homepageSettingisEnabled;
```
I like that you're trying to improve components, but here they make the code less readable imo:

* Multiple returns happening in different places makes them easy to miss. Returns should ideally be grouped as closely together as possible, just like variable definitions
* I doubt the usefulness of the extra variables (with valid in their names). I need to check what they mean higher up in the code, while if their value is just used, I know immediately what they mean. They're also used in 1 place only. Variables are mostly useful if a value is reused in multiple places.
* These variable names are a bit confusing to me. 'Valid' makes me think validation is happening, we're only checking if the values are present

So I would suggest to leave it as is. (I also think it's not productive to spend more time on discussing this given the work we still need to do, so in case you don't agree with my comments, just leave your code as is or make the changes you agree with.)

lilkraftwerk replied:
Haha, guess we have completely different styles on things like this.  Nested logical operators like this:
```
  const isAllowed =
    !appConfigSettingName ||
    (!isNilOrError(appConfigSetting) ? appConfigSetting.allowed : false);
```

I find very difficult to read and reason about. Of course I can work it out, but there's a lot of logic going on here, and my personal preference is to make things a bit more verbose for clarity. In this case it's by capturing logical operations with descriptive variables earlier in the file and then using them as shorthand in the operations later. The use of a variable can capture the meaning of a logical operation in human-readable form, as I've tried to do here, even if the variable is only used once. In other words, doing that calculation further up in the file, naming it, and then using that name later, rather than doing all the logic in line as part of other calculations.

lilkraftwerk replied:
Good point on "valid" implying some sort of validation, I'll update those variables.

lilkraftwerk replied:
I've changed them to `exists` which seems more descriptive. `isNotNilOrError` would be most accurate but is a bit longer

brentguf commented on front/app/hooks/useHomepageSettingsFeatureFlag.test.ts:
```diff
@@ -0,0 +1,86 @@
+    );
+    expect(result.current).toBe(false);
+  });
+});
```
Great, thanks for adding these! Could we add 1 more test that tests events_widget for both enabled and allowed as false? So we cover all optioins.

brentguf reviewed (commented):
Would be nice if you could add 1 more test.

lilkraftwerk committed:
Rename variables

lilkraftwerk committed:
Update variable names

lilkraftwerk committed:
Add test for disabled/disallowed events widget

lilkraftwerk commented:
Added that additional test. FE tests are failing due to a typing issue in one of the CTA files, I think it can be fixed on the feature branch

brentguf reviewed (approved):
The changed variable names indeed increase readability, I can live with them 😁 thanks for the extra test.

Will fix the missing prop in 1009.
